### PR TITLE
Between organizations scenario implementation

### PR DESCRIPTION
**Pull request by DamithaSenevirathne**

### CODE REVIEW

File: src/server/JasmineGraphServer.cpp [hunk 1]
```diff
@@ -2280,6 +2347,269 @@ bool JasmineGraphServer::initiateClient(std::string host, int port, int dataPort
...
+
+    close(sockfd);
+    return 0;
+}
+
+bool JasmineGraphServer::initiateOrgServer(std::string host, int port, int dataPort,std::string trainingArgs,int iteration, string partCount) {
+
+    Utils utils;
+    bool result = true;
+    std::cout << pthread_self() << " host : " << host << " port : " << port << " DPort : " << dataPort << std::endl;
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:2431]: Use logging framework if this line has to be captured

File: src/server/JasmineGraphServer.cpp [hunk 2]
```diff
@@ -2280,6 +2347,269 @@ bool JasmineGraphServer::initiateClient(std::string host, int port, int dataPort
...
+    char data[FED_DATA_LENGTH];
+    bool loop = false;
+    socklen_t len;
+    struct sockaddr_in serv_addr;
+    struct hostent *server;
+
+    sockfd = socket(AF_INET, SOCK_STREAM, 0);
+
+    if (sockfd < 0) {
+        std::cerr << "Cannot accept connection" << std::endl;
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:2442]: Use logging framwork

File: src/server/JasmineGraphServer.cpp [hunk 3]
```diff
@@ -2280,6 +2347,269 @@ bool JasmineGraphServer::initiateClient(std::string host, int port, int dataPort
...
+        return 0;
+    }
+
+    if (host.find('@') != std::string::npos) {
+        host = utils.split(host, '@')[1];
+    }
+
+    server = gethostbyname(host.c_str());
+    if (server == NULL) {
+        std::cerr << "ERROR, no host named " << server << std::endl;
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:2452]: Use logging framework

File: src/server/JasmineGraphServer.cpp [hunk 4]
```diff
@@ -2280,6 +2347,269 @@ bool JasmineGraphServer::initiateClient(std::string host, int port, int dataPort
...
+    }
+
+    bzero((char *) &serv_addr, sizeof(serv_addr));
+    serv_addr.sin_family = AF_INET;
+    bcopy((char *) server->h_addr,
+          (char *) &serv_addr.sin_addr.s_addr,
+          server->h_length);
+    serv_addr.sin_port = htons(port);
+    if (connect(sockfd, (struct sockaddr *) &serv_addr, sizeof(serv_addr)) < 0) {
+        std::cerr << "ERROR connecting" << std::endl;
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:2462]: Use logging framework

File: src/server/JasmineGraphServer.cpp [hunk 5]
```diff
@@ -2280,6 +2347,269 @@ bool JasmineGraphServer::initiateClient(std::string host, int port, int dataPort
...
+    struct sockaddr_in serv_addr;
+    struct hostent *server;
+
+    std::string weights_file = utils.getJasmineGraphProperty("org.jasminegraph.fl.weights.file");
+    std::string flagPath = utils.getJasmineGraphProperty("org.jasminegraph.fl.flag.file");
+
+    sockfd = socket(AF_INET, SOCK_STREAM, 0);
+
+    if (sockfd < 0) {
+        std::cerr << "Cannot accept connection" << std::endl;
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:2523]: Use logging framework

File: src/server/JasmineGraphServer.cpp [hunk 6]
```diff
@@ -2280,6 +2347,269 @@ bool JasmineGraphServer::initiateClient(std::string host, int port, int dataPort
...
+        return 0;
+    }
+    if (host.find('@') != std::string::npos) {
+        host = utils.split(host, '@')[1];
+    }
+    server = gethostbyname(host.c_str());
+    if (server == NULL) {
+        std::cerr << "ERROR, no host named " << server << std::endl;
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:2531]: Use logging framework

File: src/server/JasmineGraphServer.cpp [hunk 7]
```diff
@@ -2280,6 +2347,269 @@ bool JasmineGraphServer::initiateClient(std::string host, int port, int dataPort
...
+        read(sockfd, data, WEIGHTS_DATA_LENGTH);
+        std::string length = "";
+
+        for (int i=0; i< HEADER_LENGTH; i++) {
+            if (data[i] != NULL){
+                length += data[i];
+            }
+        }
+
+        std::cout << "message length :" << length << endl;
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:2564]: remove this cout

File: src/server/JasmineGraphServer.cpp [hunk 8]
```diff
@@ -2280,6 +2347,269 @@ bool JasmineGraphServer::initiateClient(std::string host, int port, int dataPort
...
+        char content[stoi(length)];
+        std::cout << "message length :" << length << endl;
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:2566]: remove this cout

File: src/util/Utils.cpp [hunk 9]
```diff
@@ -503,4 +503,51 @@ void Utils::copyToDirectory(std::string currentPath, std::string copyPath) {
...
+}
+
+void Utils::editFlagOne(std::string flagPath){
+    
+    std::string filePath = flagPath;
+    ofstream stream;
+    char flag[] = "1";
+
+    stream.open(filePath);
+    if( !stream )
```
Comment (miyurud) [inline src/util/Utils.cpp:515]: enclose this if within {}

File: src/util/Utils.cpp [hunk 10]
```diff
@@ -503,4 +503,51 @@ void Utils::copyToDirectory(std::string currentPath, std::string copyPath) {
...
+        cout << "Opening file failed" << endl;
+    stream << flag << endl;
+    if( !stream )
+        cout << "Write failed" << endl;
```
Comment (miyurud) [inline src/util/Utils.cpp:519]: enclose this if within {}

File: src/util/Utils.cpp [hunk 11]
```diff
@@ -503,4 +503,51 @@ void Utils::copyToDirectory(std::string currentPath, std::string copyPath) {
...
+}
+
+void Utils::editFlagZero(std::string flagPath){
+    
+    std::string filePath = flagPath;
+    ofstream stream;
+    char flag[] = "0";
+
+    stream.open(filePath);
+    if( !stream )
```
Comment (miyurud) [inline src/util/Utils.cpp:531]: enclose if within {}

File: src/util/Utils.cpp [hunk 12]
```diff
@@ -503,4 +503,51 @@ void Utils::copyToDirectory(std::string currentPath, std::string copyPath) {
...
+        cout << "Opening file failed" << endl;
+    stream << flag << endl;
+    if( !stream )
```
Comment (miyurud) [inline src/util/Utils.cpp:534]: enclose if within {}

**Review by miyurud — CHANGES_REQUESTED** [review]
Please address all the review comments in this PR and reply to each comment explaining how you addressed the issue.

File: conf/jasminegraph-server.properties [hunk 13]
```diff
@@ -47,9 +47,15 @@ org.jasminegraph.graphsage=/var/tmp/jasminegraph/GraphSAGE/graphsage/
...
 org.jasminegraph.fl.location=/home/ubuntu/software/jasminegraph/src-python/
 org.jasminegraph.fl.dataDir=/home/ubuntu/software/jasminegraph/src-python/data/
-org.jasminegraph.fl.weights=/home/ubuntu/software/jasminegraph/src-python//weights/
+org.jasminegraph.fl.weights=/home/ubuntu/software/jasminegraph/src-python/weights/
 org.jasminegraph.fl_clients=2
 org.jasminegraph.fl.epochs=4
+org.jasminegraph.fl.rounds=4
+org.jasminegraph.fl.org.port=5050
+org.jasminegraph.fl.num.orgs=1
+org.jasminegraph.fl.weights.file=/home/ubuntu/software/jasminegraph/src-python/temp/weights.txt
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply conf/jasminegraph-server.properties:56]: Yes, the Weight file is initialized by the jasmineGraph.

File: conf/jasminegraph-server.properties [hunk 14]
```diff
@@ -47,9 +47,15 @@ org.jasminegraph.graphsage=/var/tmp/jasminegraph/GraphSAGE/graphsage/
...
+org.jasminegraph.fl.flag.file=/home/ubuntu/software/jasminegraph/src-python/temp/flag.txt
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply conf/jasminegraph-server.properties:57]: currently, there are two values(0/1). These values are handled by the python implementation and jasmineGraph implementation.

File: conf/jasminegraph-server.properties (hunk 13 above)
[the remark this answers is not in the document]
Comment (miyurud) [reply conf/jasminegraph-server.properties:56]: Ok

File: conf/jasminegraph-server.properties (hunk 14 above)
[the remark this answers is not in the document]
Comment (miyurud) [reply conf/jasminegraph-server.properties:57]: Ok, then user does not need to worry about setting these values right?

[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply conf/jasminegraph-server.properties:57]: Yes

File: src/server/JasmineGraphInstanceService.cpp [hunk 15]
```diff
@@ -648,6 +648,54 @@ void *instanceservicesession(void *dummyPt) {
             std::thread workerThread = std::thread(&JasmineGraphInstanceService::initServer, trainData);
             workerThread.join();
 
+        } else if (line.compare(JasmineGraphInstanceProtocol::INITIATE_ORG_SERVER) == 0) {
+            
+            instance_logger.log("Received from: " + JasmineGraphInstanceProtocol::INITIATE_ORG_SERVER, "info");
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphInstanceService.cpp]: Fixed in 7d60572fa169b39c544c21a71f247f078a73d53a

File: src/server/JasmineGraphInstanceService.cpp [hunk 16]
```diff
@@ -1825,14 +1873,64 @@ void JasmineGraphInstanceService::initServer(string trainData){
     std::transform(trainargs.begin(), trainargs.end(), std::back_inserter(vc), converter);
 
     std::string path = "cd " + utils.getJasmineGraphProperty("org.jasminegraph.fl.location") + " && ";
-    std::string command = path + "python3 fl_server.py "+ utils.getJasmineGraphProperty("org.jasminegraph.fl.weights") + " " 
+    std::string command = path + "python fl_server.py "+ utils.getJasmineGraphProperty("org.jasminegraph.fl.weights") + " "
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphInstanceService.cpp]: Fixed in 7d60572fa169b39c544c21a71f247f078a73d53a

File: src/server/JasmineGraphInstanceService.cpp [hunk 17]
```diff
@@ -1846,15 +1944,17 @@ void JasmineGraphInstanceService::initClient(string trainData){
         }
     }
 
+
     std::vector<char *> vc;
     std::transform(trainargs.begin(), trainargs.end(), std::back_inserter(vc), converter);
 
     std::string path = "cd " + utils.getJasmineGraphProperty("org.jasminegraph.fl.location") + " && ";
-    std::string command = path + "python3 fl_client.py "+ utils.getJasmineGraphProperty("org.jasminegraph.fl.weights") + " " 
+    std::string command = path + "python fl_client.py "+ utils.getJasmineGraphProperty("org.jasminegraph.fl.weights") + " "
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphInstanceService.cpp]: Fixed in 7d60572fa169b39c544c21a71f247f078a73d53a

File: src/server/JasmineGraphInstanceService.cpp [hunk 18]
```diff
@@ -1825,14 +1873,64 @@ void JasmineGraphInstanceService::initServer(string trainData){
...
+            graphID = trainargs[i + 1];
+            break;
+        }
+    }
+
+    std::vector<char *> vc;
+    std::transform(trainargs.begin(), trainargs.end(), std::back_inserter(vc), converter);
+
+    std::string path = "cd " + utils.getJasmineGraphProperty("org.jasminegraph.fl.location") + " && ";
+    std::string command = path + "python org_server.py " + graphID+ " " + utils.getJasmineGraphProperty("org.jasminegraph.fl_clients")
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphInstanceService.cpp]: Fixed in 7d60572fa169b39c544c21a71f247f078a73d53a

File: src/server/JasmineGraphInstanceService.cpp [hunk 19]
```diff
@@ -1825,14 +1873,64 @@ void JasmineGraphInstanceService::initServer(string trainData){
...
+            graphID = trainargs[i + 1];
+            break;
+        }
+    }
+
+    std::vector<char *> vc;
+    std::transform(trainargs.begin(), trainargs.end(), std::back_inserter(vc), converter);
+
+    std::string path = "cd " + utils.getJasmineGraphProperty("org.jasminegraph.fl.location") + " && ";
+    std::string command = path + "python org_agg.py "+ " "
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphInstanceService.cpp]: Fixed in 7d60572fa169b39c544c21a71f247f078a73d53a

File: src/util/Utils.cpp (hunk 9 above)
Comment (DamithaSenevirathne) [reply src/util/Utils.cpp]: Fixed in 083074a6c0c0b2f75d2b6a0e81105d64b0a6784d

File: src/util/Utils.cpp (hunk 10 above)
Comment (DamithaSenevirathne) [reply src/util/Utils.cpp]: Fixed in 083074a6c0c0b2f75d2b6a0e81105d64b0a6784d

File: src/util/Utils.cpp (hunk 11 above)
Comment (DamithaSenevirathne) [reply src/util/Utils.cpp]: Fixed in 083074a6c0c0b2f75d2b6a0e81105d64b0a6784d

File: src/util/Utils.cpp (hunk 12 above)
Comment (DamithaSenevirathne) [reply src/util/Utils.cpp]: Fixed in 083074a6c0c0b2f75d2b6a0e81105d64b0a6784d

File: src/server/JasmineGraphServer.cpp [hunk 20]
```diff
@@ -1998,6 +1998,73 @@ void JasmineGraphServer::initiateCommunication(std::string graphID, std::string
...
+                                        stoi(workerVector[0].dataPort),trainingArgs,fl_clients, "1");
+        orgAggThread[0].join();
+        sleep(5);
+
+    }
+
+    std::thread *communicationThread = new std::thread[1];
+    communicationThread[0] = std::thread(receiveGlobalWeights,"localhost", 5000, "0", 1, "1");
+    sleep(5);
+    std::cout << "communication thread initiated" << std::endl;
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphServer.cpp]: Fixed in 3838b63e0c10352eee23ceff37cafc3492b3c14e

File: src/server/JasmineGraphServer.cpp [hunk 21]
```diff
@@ -2280,6 +2347,269 @@ bool JasmineGraphServer::initiateClient(std::string host, int port, int dataPort
     return 0;
 }
 
+bool JasmineGraphServer::initiateAggregator(std::string host, int port, int dataPort,std::string trainingArgs,int iteration, string partCount) {
+    Utils utils;
+    bool result = true;
+    std::cout << pthread_self() << " host : " << host << " port : " << port << " DPort : " << dataPort << std::endl;
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphServer.cpp]: Fixed in 3838b63e0c10352eee23ceff37cafc3492b3c14e

File: src/server/JasmineGraphServer.cpp [hunk 22]
```diff
@@ -2280,6 +2347,269 @@ bool JasmineGraphServer::initiateClient(std::string host, int port, int dataPort
...
+    char data[FED_DATA_LENGTH];
+    bool loop = false;
+    socklen_t len;
+    struct sockaddr_in serv_addr;
+    struct hostent *server;
+
+    sockfd = socket(AF_INET, SOCK_STREAM, 0);
+
+    if (sockfd < 0) {
+        std::cerr << "Cannot accept connection" << std::endl;
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphServer.cpp]: Fixed in 3838b63e0c10352eee23ceff37cafc3492b3c14e

File: src/server/JasmineGraphServer.cpp [hunk 23]
```diff
@@ -2280,6 +2347,269 @@ bool JasmineGraphServer::initiateClient(std::string host, int port, int dataPort
...
+        return 0;
+    }
+
+    if (host.find('@') != std::string::npos) {
+        host = utils.split(host, '@')[1];
+    }
+
+    server = gethostbyname(host.c_str());
+    if (server == NULL) {
+        std::cerr << "ERROR, no host named " << server << std::endl;
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphServer.cpp]: Fixed in 3838b63e0c10352eee23ceff37cafc3492b3c14e

File: src/server/JasmineGraphServer.cpp [hunk 24]
```diff
@@ -2280,6 +2347,269 @@ bool JasmineGraphServer::initiateClient(std::string host, int port, int dataPort
...
+    }
+
+    bzero((char *) &serv_addr, sizeof(serv_addr));
+    serv_addr.sin_family = AF_INET;
+    bcopy((char *) server->h_addr,
+          (char *) &serv_addr.sin_addr.s_addr,
+          server->h_length);
+    serv_addr.sin_port = htons(port);
+    if (connect(sockfd, (struct sockaddr *) &serv_addr, sizeof(serv_addr)) < 0) {
+        std::cerr << "ERROR connecting" << std::endl;
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphServer.cpp]: Fixed in 3838b63e0c10352eee23ceff37cafc3492b3c14e

File: src/server/JasmineGraphServer.cpp [hunk 25]
```diff
@@ -2280,6 +2347,269 @@ bool JasmineGraphServer::initiateClient(std::string host, int port, int dataPort
...
+        server_logger.log("Sent : " + server_host, "info");
+
+        write(sockfd, JasmineGraphInstanceProtocol::INITIATE_AGG.c_str(),
+              JasmineGraphInstanceProtocol::INITIATE_AGG.size());
+        server_logger.log("Sent : " + JasmineGraphInstanceProtocol::INITIATE_AGG, "info");
+        bzero(data, FED_DATA_LENGTH);
+        read(sockfd, data, FED_DATA_LENGTH);
+        response = (data);
+        response = utils.trim_copy(response, " \f\n\r\t\v");
+        std::cout << response << std::endl;
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphServer.cpp]: Replaced in  3838b63e0c10352eee23ceff37cafc3492b3c14e

File: src/server/JasmineGraphServer.cpp (hunk 1 above)
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphServer.cpp]: Fixed in 3838b63e0c10352eee23ceff37cafc3492b3c14e

File: src/server/JasmineGraphServer.cpp (hunk 2 above)
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphServer.cpp]: Fixed in 3838b63e0c10352eee23ceff37cafc3492b3c14e

File: src/server/JasmineGraphServer.cpp (hunk 3 above)
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphServer.cpp]: Fixed in 3838b63e0c10352eee23ceff37cafc3492b3c14e

File: src/server/JasmineGraphServer.cpp (hunk 4 above)
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphServer.cpp]: Fixed in 3838b63e0c10352eee23ceff37cafc3492b3c14e

File: src/server/JasmineGraphServer.cpp (hunk 5 above)
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphServer.cpp]: Fixed in 3838b63e0c10352eee23ceff37cafc3492b3c14e

File: src/server/JasmineGraphServer.cpp (hunk 6 above)
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphServer.cpp]: Fixed in 3838b63e0c10352eee23ceff37cafc3492b3c14e

File: src/server/JasmineGraphServer.cpp (hunk 7 above)
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphServer.cpp]: Removed in  3838b63e0c10352eee23ceff37cafc3492b3c14e

File: src/server/JasmineGraphServer.cpp (hunk 8 above)
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphServer.cpp]: Removed in 3838b63e0c10352eee23ceff37cafc3492b3c14e

File: src/server/JasmineGraphServer.cpp [hunk 26]
```diff
@@ -1998,6 +1998,73 @@ void JasmineGraphServer::initiateCommunication(std::string graphID, std::string
 
 }
 
+void JasmineGraphServer::initiateOrgCommunication(std::string graphID, std::string trainingArgs, SQLiteDBInterface sqlite) {
+
+    Utils utils;
+    int fl_clients = stoi(utils.getJasmineGraphProperty("org.jasminegraph.fl_clients"));
+        std::string flagPath = utils.getJasmineGraphProperty("org.jasminegraph.fl.flag.file");
+    int threadLimit = fl_clients+1;
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphServer.cpp]: Chenged in e8734ce1f1ee037f4cfb8ac6df02d026253dc577

File: src/server/JasmineGraphInstanceService.cpp [hunk 27]
```diff
@@ -1825,14 +1873,64 @@ void JasmineGraphInstanceService::initServer(string trainData){
...
+    std::string path = "cd " + utils.getJasmineGraphProperty("org.jasminegraph.fl.location") + " && ";
+    std::string command = path + "python org_server.py " + graphID+ " " + utils.getJasmineGraphProperty("org.jasminegraph.fl_clients")
+                                    + " " + utils.getJasmineGraphProperty("org.jasminegraph.fl.epochs") 
+                                    +" localhost 5050 > org_server_logs.txt";
+    popen(command.c_str(), "r");    
+}
+
+void JasmineGraphInstanceService::initAgg(string trainData){
+    Utils utils;
+    std::cout << "initiate Organizational Aggregator" << std::endl;
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphInstanceService.cpp]: Removed in abcd7273140d0e7d1be756000b686a83dc4a8d9f

File: src/server/JasmineGraphInstanceService.cpp [hunk 28]
```diff
@@ -1825,14 +1873,64 @@ void JasmineGraphInstanceService::initServer(string trainData){
...
                                 + utils.getJasmineGraphProperty("org.jasminegraph.fl.dataDir") 
                                 + " " + utils.getJasmineGraphProperty("org.jasminegraph.fl.dataDir")+ " "+ graphID + " 0 "
                                 + utils.getJasmineGraphProperty("org.jasminegraph.fl_clients")
                                 + " " + utils.getJasmineGraphProperty("org.jasminegraph.fl.epochs") +" localhost 5000 > server_logs.txt";
     popen(command.c_str(), "r");    
 }
 
+void JasmineGraphInstanceService::initOrgServer(string trainData){
+    Utils utils;
+    std::cout << "inside server" << std::endl;
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphInstanceService.cpp]: Removed in b2aa951ed8b33d1ddd1b42fab17a6f856f99ea27

File: src/server/JasmineGraphServer.cpp [hunk 29]
```diff
@@ -1998,6 +1998,73 @@ void JasmineGraphServer::initiateCommunication(std::string graphID, std::string
...
+        workerInstance = workerVector[i];
+
+        int serverPort = stoi(workerInstance.port);
+        int serverDataPort = stoi(workerInstance.dataPort);
+
+        if (i==0) {
+
+            workerThreads[threadID] = std::thread(initiateOrgServer,"localhost", serverPort, 
+                                                    serverDataPort,trainingArgs,fl_clients, to_string(i));
+            sleep(5);
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphServer.cpp]: Removed in e8734ce1f1ee037f4cfb8ac6df02d026253dc577

File: src/server/JasmineGraphServer.cpp [hunk 30]
```diff
@@ -1998,6 +1998,73 @@ void JasmineGraphServer::initiateCommunication(std::string graphID, std::string
...
+    }
+   
+    workerThreads[0].join();
+    sleep(5);
+    communicationThread[0].join();
+
+    for (int threadCount = 1; threadCount < threadLimit; threadCount++) {
+        workerThreads[threadCount].join();
+        std::cout << "Thread " << threadCount << " joined" << std::endl;
+        sleep(10);
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply src/server/JasmineGraphServer.cpp]: Removed in e8734ce1f1ee037f4cfb8ac6df02d026253dc577

File: conf/jasminegraph-server.properties (hunk 14 above)
[the remark this answers is not in the document]
Comment (miyurud) [reply conf/jasminegraph-server.properties:57]: Ok, understood.

File: src/server/JasmineGraphInstanceService.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp]: Done

File: src/server/JasmineGraphInstanceService.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp]: Done

File: src/server/JasmineGraphInstanceService.cpp (hunk 28 above)
[the remark this answers is not in the document]
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp]: Done

File: src/server/JasmineGraphServer.cpp (hunk 8 above)
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Done

File: src/server/JasmineGraphInstanceService.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp]: Done

File: src/server/JasmineGraphInstanceService.cpp (hunk 27 above)
[the remark this answers is not in the document]
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp]: Done

File: src/server/JasmineGraphInstanceService.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp]: Done

File: src/server/JasmineGraphInstanceService.cpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp]: Done

File: src/server/JasmineGraphServer.cpp (hunk 26 above)
[the remark this answers is not in the document]
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Done

File: src/server/JasmineGraphServer.cpp [hunk 31]
```diff
@@ -1998,6 +1998,73 @@ void JasmineGraphServer::initiateCommunication(std::string graphID, std::string
...
+
+    vector<Utils::worker> workerVector = utils.getWorkerList(sqlite);
+
+    if (utils.getJasmineGraphProperty("org.jasminegraph.fl.aggregator") == "true"){
+   
+        std::thread *orgAggThread = new std::thread[1];
+        orgAggThread[0] = std::thread(initiateAggregator,"localhost",  stoi(workerVector[0].port), 
+                                        stoi(workerVector[0].dataPort),trainingArgs,fl_clients, "1");
+        orgAggThread[0].join();
+        sleep(5);
```
[the remark this answers is not in the document]
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Done

File: src/server/JasmineGraphServer.cpp (hunk 29 above)
[the remark this answers is not in the document]
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Done

File: src/server/JasmineGraphServer.cpp (hunk 30 above)
[the remark this answers is not in the document]
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Done

File: src/server/JasmineGraphServer.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Done

File: src/server/JasmineGraphServer.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Done

File: src/server/JasmineGraphServer.cpp (hunk 24 above)
[the remark this answers is not in the document]
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Done

File: src/server/JasmineGraphServer.cpp (hunk 25 above)
[the remark this answers is not in the document]
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: You have actually removed the line. Thats fine.
Done

File: src/server/JasmineGraphServer.cpp (hunk 1 above)
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Done

File: src/server/JasmineGraphServer.cpp (hunk 2 above)
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Done

File: src/server/JasmineGraphServer.cpp (hunk 3 above)
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Done

File: src/server/JasmineGraphServer.cpp (hunk 4 above)
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Done

File: src/server/JasmineGraphServer.cpp (hunk 6 above)
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Done

File: src/server/JasmineGraphServer.cpp (hunk 5 above)
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Done

File: src/server/JasmineGraphServer.cpp (hunk 7 above)
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Done

File: src/util/Utils.cpp (hunk 9 above)
Comment (miyurud) [reply src/util/Utils.cpp]: Done

File: src/util/Utils.cpp (hunk 10 above)
Comment (miyurud) [reply src/util/Utils.cpp]: Done

File: src/util/Utils.cpp (hunk 11 above)
Comment (miyurud) [reply src/util/Utils.cpp]: Done

File: src/util/Utils.cpp (hunk 12 above)
Comment (miyurud) [reply src/util/Utils.cpp]: Done

**Pull request merged**